core/VectorState.hpp

```diff
@@ -34,85 +80,411 @@ namespace atlas
 
         inline void setVSTART(uint8_t value) { vstart_ = value; }
 
-        template <typename VLEN> inline uint8_t getVLMAX()
+        // helper functions
+
+        inline uint8_t getVLMAX() const { return getVLEN() / 8 * getLMUL() / getSEW(); }
+
+        template <typename XLEN> XLEN vsetAVL(AtlasState* state, bool set_max, XLEN avl = 0)
         {
-            return sizeof(VLEN) * lmul_ / (sew_ == 0 ? 1 : sew_);
+            uint8_t vl = set_max ? getVLMAX() : std::min<uint8_t>(getVLMAX(), avl);
+            setVL(vl);
+            WRITE_CSR_REG<XLEN>(state, VL, vl);
+            return vl;
         }
 
         template <typename XLEN> void vsetVTYPE(AtlasState* state, XLEN vtype)
         {
             WRITE_CSR_REG<XLEN>(state, VTYPE, vtype);
-            sew_ = 8u << READ_CSR_FIELD<XLEN>(state, VTYPE, "vsew");
-            uint8_t vlmul_ = READ_CSR_FIELD<XLEN>(state, VTYPE, "vlmul");
-            switch (vlmul_)
-            {
-                case 0b101:
-                    lmul_ = 1;
-                    break;
-                case 0b110:
-                    lmul_ = 2;
-                    break;
-                case 0b111:
-                    lmul_ = 4;
-                    break;
-                case 0b000:
-                    lmul_ = 8;
-                    break;
-                case 0b001:
-                    lmul_ = 16;
-                    break;
-                case 0b010:
-                    lmul_ = 32;
-                    break;
-                case 0b011:
-                    lmul_ = 64;
-                    break;
-                default:
-                    break;
-            }
-            vta_ = READ_CSR_FIELD<XLEN>(state, VTYPE, "vta");
-            vma_ = READ_CSR_FIELD<XLEN>(state, VTYPE, "vma");
-        }
+            uint8_t vlmul = READ_CSR_FIELD<XLEN>(state, VTYPE, "vlmul");
 
-        template <typename XLEN, typename VLEN>
-        XLEN vsetAVL(AtlasState* state, bool set_max, XLEN avl = 0)
-        {
-            vl_ = set_max ? getVLMAX<VLEN>() : std::min<uint8_t>(getVLMAX<VLEN>(), avl);
-            WRITE_CSR_REG<XLEN>(state, VL, vl_);
-            return vl_;
+            static const uint8_t lmul_table[8] = {
+                8,  // 000
+                16, // 001
+                32, // 010
+                64, // 011
+                0,  // 100 (invalid)
+                1,  // 101
+                2,  // 110
+                4   // 111
+            };
+
+            uint8_t lmul = lmul_table[vlmul & 0b111];
```

Comment:
Variables that are assigned during creating and not assigned again should be const